### DISCUSSION TITLE
KTXLoader VS HDRLoader

### DISCUSSION THREAD

**Discussion by ThomasGorisse** [google/filament#4562] [Q&A]
Hi,

I'm currently working on Scenform Maintained indirect lights and I'm a little confused between KTXLoader and HDRLoader.
I saw that you introduced the HDRLoader on the last Android GLTFViewer Sample but also use the KTXLoader at the same time.
- https://github.com/google/filament/blob/9f5d6e79da2429ecd5bc6dd66c2531cb50798c29/android/samples/sample-gltf-viewer/src/main/java/com/google/android/filament/gltf/MainActivity.kt#L178
- https://github.com/google/filament/blob/9f5d6e79da2429ecd5bc6dd66c2531cb50798c29/android/samples/sample-gltf-viewer/src/main/java/com/google/android/filament/gltf/MainActivity.kt#L129

What are the main differences between them ?

I need to apply the Spherical Harmonics coefficient from ARCore and since i can retrieve them from the .ktx using KTXLoader, I'm able to create a new IndirectLight each frame with the factored irradiance.

If there is no difference between them, can I retrieve the reflections texture from the KTXLoader ?

```kotlin
IBLPrefilterContext.SpecularFilter(context)
  .run(skyboxTexture)?.let { reflectionsTexture ->
      imageBasedLight(
          IndirectLight.Builder()
              .reflections(reflectionsTexture)
              .intensity(30000.0f)
              .build(engine)
      )
  } ?: error(Exception("Could not decode reflections texture"))
```

Thanks

**Comment by romainguy** [comment]
`HDRLoader` loads .hdr file (Radiance format) and `KTXLoader` loads .ktx files. The KTX files we use have been pre-processed by `cmgen` but that's not the case of HDR files (equirectilinear source images you normally feed `cmgen`).

The sample uses the two file formats for two different reasons:
- The sample comes bundled with a pre-processed .ktx file, used as the default indirect light
- Using our [remote control web UI](https://google.github.io/filament/remote/) you can change the IBL by just drag and dropping a .hdr file, which will then be sent to the running app on the phone to be loaded and pre-filtered (see below)

ARCore should be giving you raw image data, so neither `KTXLoader` nor `HDRLoader` is what you want. `IBLPrefilterContext` will take a decoded environment map (as a texture) and pre-filter it the way `cmgen` does, except on the GPU. It's what you want to use with what ARCore gives you. If you've generated a .ktx with `cmgen`, you don't need to use `IBLPrefilterContext`.

**Reply by ThomasGorisse** [reply]
Thanks for the clarification.

> ARCore should be giving you raw image data, so neither KTXLoader nor HDRLoader is what you want.

Actually, I have to use it for this 3 of 4 usage cases:
- Non AR usage default environment
- AR without what ARCore calls HDR lightning available/enable:
    - Indirect light with static "fake" reflections from .hdr, AR intensity,... 
    - Directional sunlight with AR rotation, position,...
- AR with ARCore HDR Lightning but static reflexions since the ARCore ones are too low res for now.
